src/protocols/secure_channel/CASESession.cpp

```diff
@@ -226,6 +240,12 @@ class CASESession::WorkHelper
         auto status = DeviceLayer::PlatformMgr().ScheduleWork(AfterWorkHandler, reinterpret_cast<intptr_t>(helper));
         if (status != CHIP_NO_ERROR)
         {
+            // We failed to schedule after work callback, so setting mScheduleAfterWorkFailed flag to true
+            // This can be checked from foreground thread and after work callback can be retried
+            ChipLogError(SecureChannel, "Failed to Schedule the AfterWorkCallback on foreground thread");
+            helper->mScheduleAfterWorkFailed.store(true);
+            helper->mStatus = status;
```

Comment:
This line needs to be before we set mScheduleAfterWorkFailed.  We're on a background thread here, and once we set mScheduleAfterWorkFailed to true we can get our DoAfterWork called on the other thread.  If that happens before mStatus gets set, that would be bad.  So need to set mStatus first, then do the synchronization point on the atomic boolean.

Reply:
Thanks for pointing out and explanation, moved above.